### PR TITLE
suggest list type columns for annotations with multiple values (PLFM-6013)

### PR DESCRIPTION
**Pull request by zimingd**

### CODE REVIEW

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java [hunk 1]
```diff
@@ -885,9 +886,19 @@ public ColumnAggregation mapRow(ResultSet rs, int rowNum)
 				ColumnModel model = new ColumnModel();
 				model.setName(aggregation.getColumnName());
 				ColumnType type = AnnotationType.valueOf(typeString).getColumnType();
+
+				//check if a LIST columnType needs to be used
+				if(aggregation.getListSize() > 1){
+					try {
+						type = ColumnTypeListMappings.forNonListType(type).getListType();
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java]: optional might work better than illegal arg for getListType();

File: lib/models/src/main/java/org/sagebionetworks/repo/model/table/TableConstants.java [hunk 2]
```diff
@@ -289,10 +292,12 @@ public static Long getReservedColumnId(String columnName) {
...
 	
-	public static final String SELECT_DISTINCT_ANNOTATION_COLUMNS_TEMPLATE = "SELECT A."
-			+ ANNOTATION_REPLICATION_COL_KEY + ", GROUP_CONCAT(DISTINCT A."
-			+ ANNOTATION_REPLICATION_COL_TYPE + "), MAX(LENGTH(A."
-			+ ANNOTATION_REPLICATION_COL_STRING_VALUE + "))" + " FROM "
+	public static final String SELECT_DISTINCT_ANNOTATION_COLUMNS_TEMPLATE = "SELECT "
+			+"A." + ANNOTATION_REPLICATION_COL_KEY
+			+", GROUP_CONCAT(DISTINCT A." + ANNOTATION_REPLICATION_COL_TYPE + ")"
+			+", MAX("+ANNOTATION_REPLICATION_COL_MAX_STRING_LENGTH+")"
+			+", JSON_LENGTH("+ANNOTATION_REPLICATION_COL_STRING_LIST_VALUE+")"
```
Comment (john-hill) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/table/TableConstants.java]: max(json_lenght())
Need  a test that would show this returns the wrong value.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java [hunk 3]
```diff
@@ -885,9 +886,19 @@ public ColumnAggregation mapRow(ResultSet rs, int rowNum)
...
+					} catch (IllegalArgumentException e){
+						//do nothing because a list type mapping does not exist
+					}
+				}
+
 				model.setColumnType(type);
-				if(ColumnType.STRING == type) {
-					model.setMaximumSize(aggregation.getMaxSize());
+				if(ColumnType.STRING == type || ColumnType.STRING_LIST==type) {
+					model.setMaximumSize(aggregation.getMaxStringElementSize());
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java:901]: maxListSize

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLUtilsTest.java [hunk 4]
```diff
@@ -2625,6 +2628,8 @@ public void testWriteAnnotationDtoToPreparedStatementLongList() throws SQLExcept
 		verify(mockPreparedStatement).setString(9, "[\"123\",\"456\",\"789\"]");
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLUtilsTest.java]: all 3s?

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/TableIndexDAOImplTest.java [hunk 5]
```diff
@@ -1839,28 +1866,84 @@ public void testCaseSensitiveAnnotationNamesPLFM_5449() {
...
+		// delete all data
+		tableIndexDAO.deleteEntityData(Lists.newArrayList(2L,3L));
+
+
+		String annoKey = "myAnnotation";
+
+		// setup some hierarchy.
+		EntityDTO file1 = createEntityDTO(2L, EntityType.file, 0);
+		file1.setParentId(333L);
+
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/TableIndexDAOImplTest.java:1882]: first annotation should not be a list.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java (hunk 1 above)
Comment (zimingd) [reply lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAOImpl.java]: Other parts of code is relying on it throwing an exception being thrown if unexpected usage of  conversion to list type occurs

**Review by john-hill — APPROVED** [review]

**Pull request merged**